File: gathering-api/src/main/java/gg/warcraft/gathering/api/gatherable/Gatherable.java
package gg.warcraft.gathering.api.gatherable;

import gg.warcraft.monolith.api.util.Duration;
import gg.warcraft.monolith.api.world.item.Item;

import java.util.List;

/**
 * A Gatherable represents a block or entity that can be harvested for resources such as a STONE block in a mine, a
 * WOOD block in a forest, or a PIG on a farm. Gatherables respawn after a cooldown.
 */
public interface Gatherable {

    /**
     * Generates a list of drops for this gatherable. This method could return different items based on chance rather
     * than the same ones every time and as such will be called whenever this gatherable is gathered.
     *
     * @return the generated drops
     */
    List<Item> generateDrops();

    /**
     * Generates a cooldown for this gatherable. This method could return different durations based on chance rather
     * than the same one every time and as such will be called whenever this gatherable is gathered.
     *
     * @return the duration after which this gatherable will respawn
     */
    Duration generateCooldown();
}

File: gathering/src/main/scala/gg/warcraft/gathering/gatherable/Gatherable.scala
package gg.warcraft.gathering.gatherable

import gg.warcraft.monolith.api.entity.{Entity, EntityType}
import gg.warcraft.monolith.api.world.block.{Block, BlockTypeVariantOrState}
import gg.warcraft.monolith.api.world.item.ItemTypeOrVariant

trait Gatherable {
  val dropData: ItemTypeOrVariant
  val dropName: String
  val cooldown: Int
  val cooldownDelta: Int
}

case class BlockGatherable(
    blockData: BlockTypeVariantOrState,
    dropData: ItemTypeOrVariant,
    dropName: String,
    cooldown: Int,
    cooldownDelta: Int,
    cooldownData: BlockTypeVariantOrState
) extends Gatherable {
  def matches(block: Block): Boolean =
    block.hasData(blockData)
}

case class EntityGatherable(
    entityType: EntityType,
    entityCount: Int,
    dropData: ItemTypeOrVariant,
    dropName: String,
    cooldown: Int,
    cooldownDelta: Int
) extends Gatherable {
  def matches(entity: Entity): Boolean =
    entity.getType == entityType
}
